service/consul/instancer.go

```diff
@@ -136,29 +145,32 @@ func (i *instancer) getInstances(lastIndex uint64, stop <-chan struct{}) ([]stri
 	result := make(chan response, 1)
 
 	go func() {
-		var queryOptions api.QueryOptions = i.queryOptions
-		queryOptions.WaitIndex = lastIndex
-		entries, meta, err := i.client.Service(i.service, i.tag, i.passingOnly, &queryOptions)
-		if err != nil {
-			result <- response{err: err}
-			return
-		}
+		var (
+			queryOptions api.QueryOptions = i.queryOptions
+			entries      []*api.ServiceEntry
+			meta         *api.QueryMeta
+			resp         response
+		)
 
-		if len(i.filterTags) > 0 {
-			entries = filterEntries(entries, i.filterTags)
-		}
-
-		// see: https://www.consul.io/api-docs/features/blocking#implementation-details
-		if meta == nil || meta.LastIndex < lastIndex {
-			lastIndex = 0
-		} else {
-			lastIndex = meta.LastIndex
+		queryOptions.WaitIndex = lastIndex
+		entries, meta, resp.err = i.client.Service(i.service, i.tag, i.passingOnly, &queryOptions)
+		if resp.err == nil {
+			// see: https://www.consul.io/api-docs/features/blocking#implementation-details
+			if meta == nil || meta.LastIndex < lastIndex {
+				resp.err = errIndexUnderflow
+			} else if meta.LastIndex == 0 {
+				resp.err = errIndexZero
```

Comment:
and setting the index to 1 in this case:
```
resp.index = 1
```
"After the initial request (or a reset as above) the X-Consul-Index returned should always be greater than zero. [...] It is always safe to use an index of 1 to wait for updates when the data being requested doesn't exist yet, so clients should sanity check that their index is at least 1 after each blocking response is handled to be sure they actually block on the next request."